Handle multi point hold notes in editor

maxrchung opened the pull request:
Closes #520 

A few miscellaneous things I touched that are somewhat related:
* Fixed visual tests looking super stretched wide by fitting all the contents into a square container
* Made hold note anchors and anchor path a fixed width so it doesn't shift when the camera zooms in/out
* Hold note end coordinate score calculation

maxrchung commented on S2VX.Game.Tests/UnitTests/S2VXStoryTests.cs:
```diff
@@ -0,0 +1,53 @@
+using System.Linq;
+
+namespace S2VX.Game.Tests.UnitTests {
+    public static class S2VXStoryTests {
```
See this section in our wiki for class vs function structure: https://github.com/maxrchung/S2VX/wiki/Testing#headlesstests. It's a bit out of date since at the time we didn't separate our tests as Unit, Headless, and Visual tests. "Headless" in the wiki's case refers to Unit tests.

Basically, any time there's an opportunity where you have common set up code and multiple asserts on that set up, I prefer the class structure. I think it better groups related test cases together and provides a common `[SetUp]` block for the test cases. Of course, the alternative is just using functions and helper functions, but I feel that having classes is nicer for encapsulating a set of assertions together.

Also as a note, we actually can't use test classes like this for our TestScene tests because osu!framework's TestScene detection is fairly primitive and can't detect nested test cases correctly.

`[TestFixture]` is not needed. NUnit is smart enough to detect a test based on the `[Test]` attribute.

Naranja-Sagged commented:
I can get full score for a slider if I hit the beginning and hold my key even if I don't move my mouse to the end part. Is this intended?

maxrchung commented:
> I can get full score for a slider if I hit the beginning and hold my key even if I don't move my mouse to the end part. Is this intended?

As far as I can tell, this logic wasn't in the code. Or maybe during one of our hold note refactors, we removed this functionality and didn't account for it. I implemented and tested this feature in this commit: https://github.com/maxrchung/S2VX/pull/523/commits/32e8d456979d24baa49b17096a52f5b3bc1fe096

However, let's probably have a group decision with @Xenocidel on this whether we want to support this. I'm under the impression that we originally wanted this in our old designs. It can be somewhat punishing though - if you miss the end note you're punished for the whole duration of the slider in score, I believe.

Xenocidel commented:
> > I can get full score for a slider if I hit the beginning and hold my key even if I don't move my mouse to the end part. Is this intended?
> 
> As far as I can tell, this logic wasn't in the code. Or maybe during one of our hold note refactors, we removed this functionality and didn't account for it. I implemented and tested this feature in this commit: [32e8d45](https://github.com/maxrchung/S2VX/commit/32e8d456979d24baa49b17096a52f5b3bc1fe096)
> 
> However, let's probably have a group decision with @Xenocidel whether we want to support this. I'm under the impression that we originally wanted this in our old designs. It can be somewhat punishing though - if you miss position the end note you're punished for the whole duration of the slider in score. But I think I prefer having this functionality - so that these notes are treated more as a slider and not just a glorified hold note.

Hm, I definitely wouldn't want to punish with a complete miss if you didn't have your cursor on the EndCoordinate at EndTime. How about you have to be over the EndCoordinate at EndTime otherwise it'd treat it as a note miss (and add MissThreshold to your score).

maxrchung commented:
> > > I can get full score for a slider if I hit the beginning and hold my key even if I don't move my mouse to the end part. Is this intended?
> > 
> > 
> > As far as I can tell, this logic wasn't in the code. Or maybe during one of our hold note refactors, we removed this functionality and didn't account for it. I implemented and tested this feature in this commit: [32e8d45](https://github.com/maxrchung/S2VX/commit/32e8d456979d24baa49b17096a52f5b3bc1fe096)
> > However, let's probably have a group decision with @Xenocidel whether we want to support this. I'm under the impression that we originally wanted this in our old designs. It can be somewhat punishing though - if you miss position the end note you're punished for the whole duration of the slider in score. But I think I prefer having this functionality - so that these notes are treated more as a slider and not just a glorified hold note.
> 
> Hm, I definitely wouldn't want to punish with a complete miss if you didn't have your cursor on the EndCoordinate at EndTime. How about you have to be over the EndCoordinate at EndTime otherwise it'd treat it as a note miss (and add MissThreshold to your score).

I have implemented according to this logic. Now when you reach the end of a hold note there are 2 score checks: 1 for the hold duration logic that we already had before, and 1 that checks if your mouse is currently over the end coordinates. In retrospect, I probably should have split this out to another checkin since it's somewhat larger than scope and less related than I realized.

I also added a line into our wiki: https://github.com/maxrchung/S2VX/wiki/Score-Calculation#hold-note-calculation
